File: Chapter 05/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter05/section_5_6_crypto_tracker/after/core/domain/PortfolioManager.java
package com.grokkingsoftwarearchitecture.chapter05.section_5_6_crypto_tracker.after.core.domain;

import com.grokkingsoftwarearchitecture.chapter05.section_5_6_crypto_tracker.after.core.ports.PriceProviderPort;

/**
 * CORE – Pure business logic.
 * No HTTP clients, no JSON parsing. This class is fully isolated.
 */
public class PortfolioManager {
    private final PriceProviderPort priceProvider;

    /**
     * Dependency Injection via Constructor. 
     */
    public PortfolioManager(PriceProviderPort priceProvider) {
        this.priceProvider = priceProvider;
    }

    public double calculateTotalValue(double btcAmount) throws Exception {
        // We just call the port. We don't care WHERE the price comes from.
        double currentPrice = priceProvider.getBitcoinPrice();

        // NEW: Log the data we just got from the "Outside World"
        System.out.println("[Core] Calculating value: " + btcAmount + " BTC @ $" + currentPrice + " per coin");

        return btcAmount * currentPrice;
    }
}

File: Chapter 05/C#/Section_5_6_Crypto_Tracker/After/Core/Domain/PortfolioManager.cs
using System.Threading.Tasks;
using Chapter05.CryptoTracker.After.Core.Ports;

namespace Chapter05.CryptoTracker.After.Core.Domain
{
    /// <summary>
    /// CORE – Pure business logic.
    /// No HTTP clients, no JSON parsing. This class is fully isolated.
    /// </summary>
    public class PortfolioManager
    {
        private readonly IPriceProviderPort _priceProvider;

        /// <summary>
        /// Dependency Injection via Constructor. 
        /// We demand a "socket" (IPriceProviderPort), but we don't care which 
        /// plug (Adapter) the Boundary Keeper decides to use!
        /// </summary>
        public PortfolioManager(IPriceProviderPort priceProvider)
        {
            _priceProvider = priceProvider;
        }

        public async Task<decimal> CalculateTotalValue(decimal btcAmount)
        {
            // We just call the port. We don't care WHERE the price comes from.
            var currentPrice = await _priceProvider.GetBitcoinPrice();
            
            // Pure math. No network dependencies here!
            return btcAmount * currentPrice;
        }
    }
}